aten/src/ATen/core/ATenDispatch.h

```diff
@@ -40,17 +42,47 @@ static inline TensorTypeId dispatchTypeId(TensorTypeSet ts) {
 
 }
 
+namespace detail {
+  struct MultiDispatchTensorTypeSet : IterArgs<MultiDispatchTensorTypeSet> {
+    TensorTypeSet ts;
+    void operator()(const at::Tensor& x) {
+      ts = ts | x.type_set();
+    }
+    void operator()(TensorOptions x) {
+      ts = ts | x.type_set();
+    }
+    void operator()(at::ArrayRef<at::Tensor> xs) {
+      for (const auto& x : xs) {
+        ts = ts | x.type_set();
+      }
+    }
+    template <typename T>
+    void operator()(const T& x) {
```

Comment:
This seems fragile. Let's assert that these are only the types we expect, either by static_assert or by only overloading the concrete types. Example: c10 takes tensor lists not as ArrayRef but as torch::List. We would probably miss that once we start using this for c10.

Reply:
This feels like a "cure is worse than the disease" situation. We support a *lot* of input types for kernels, easily dozens, and what you've proposed would effectively require us to maintain yet another list of every possible input type in the codebase, and as you've seen, we add types all the time. Whereas I feel pretty comfortable assuming that the number of Tensor-containing argument types will change very slowly in the future.

Re torch::List, I don't understand why this type exists at all in the arguments of kernels (cc @dzhulgakov, original reviewer of the PR #21164 -- I'm marked as the reviewer but I only reapproved the diff for administrative reasons). As best I can tell, `List` is an *owning* type, which means that if I have some sort of type that cannot be directly moved into List, e.g., an `initializer_list`, I have to copy the elements. But for arguments to kernels, this doesn't really make any sense: if I write:

```
convolution(x, {2, 3} /* padding */);
```

There is no reason to copy the `{2, 3}` initializer list into an owning list. I want to pass it in by reference, and that's exactly what `ArrayRef` does. Is there a use case that I'm not aware of?

Reply:
The reason for torch::List as an argument and return type was mostly consistency. You can't return ArrayRef for example, so you'd need different types if you take a list as argument vs when you return it. Also, ArrayRef doesn't work with book, so you'd need a different non-ArrayRef type for BoolList arguments. See the whole reasoning here: https://quip.com/UWUpAoaPsipL

Reply:
About the original point, you got me convinced that we shouldn't introduce a new list of supported types here. We'll hopefully notice if something is missing.

Btw talking of, donyou have a plan on how to handle Dicts with Tensor values? Nested things like List of List of Dict of ... ?

Reply:
> The reason for torch::List as an argument and return type was mostly consistency

Foolish consistency? I mean, I guess if people write kernels that take in List as an argument, I won't stop them. But it's a sort of fact of life in C++ that argument types and return types are going to look different. The best practice is going to be to use ArrayRef on input. The consistency argument also feels even weaker because List is our own home grown type and so you have to tell the user it exists.

I don't find the bool case particularly convincing either. We're not building a general purpose operator library: we're building a tensor library. I have not seen a single compelling use case for passing a variable-length list of booleans as an argument to an operator. (`torch.tensor` doesn't count; you don't want to bind that in native anyway.)

> Btw talking of missing types, do you have a plan on how to handle Dicts with Tensor values? Nested things like List of List of Dict of ... ?

We've got to draw the line somewhere. I'm pretty OK with saying, "Only Tensor, ArrayRef<Tensor>, TensorOptions" participate in multiple dispatch. We have similar limitations with autograd tracking for user defined functions: if you smuggle a Variable in, inside some complex struct, it's not going to be counted as an input for the purposes of the autograd function.

Reply:
so you can't use lazy tensors in Dicts for example?